crates/sui-single-node-benchmark/src/mock_consensus.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use prometheus::IntCounter;
+use std::sync::Arc;
+use sui_core::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use sui_core::authority::AuthorityState;
+use sui_core::checkpoints::CheckpointServiceNoop;
+use sui_core::consensus_adapter::SubmitToConsensus;
+use sui_core::consensus_handler::SequencedConsensusTransaction;
+use sui_types::error::SuiResult;
+use sui_types::messages_consensus::ConsensusTransaction;
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+
+pub(crate) struct MockConsensusClient {
+    tx_sender: mpsc::Sender<ConsensusTransaction>,
+    _consensus_handle: JoinHandle<()>,
+}
+
+pub(crate) enum ConsensusMode {
+    // ConsensusClient does absolutely nothing when receiving a transaction
+    Noop,
+    // ConsensusClient directly sequences the transaction into the store.
+    DirectSequencing,
+}
+
+impl MockConsensusClient {
+    pub(crate) fn new(validator: Arc<AuthorityState>, consensus_mode: ConsensusMode) -> Self {
+        let (tx_sender, tx_receiver) = mpsc::channel(1000000);
+        let _consensus_handle = Self::run(validator, tx_receiver, consensus_mode);
+        Self {
+            tx_sender,
+            _consensus_handle,
+        }
+    }
+
+    pub(crate) fn run(
+        validator: Arc<AuthorityState>,
+        tx_receiver: mpsc::Receiver<ConsensusTransaction>,
+        consensus_mode: ConsensusMode,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move { Self::run_impl(validator, tx_receiver, consensus_mode).await })
+    }
+
+    async fn run_impl(
+        validator: Arc<AuthorityState>,
+        mut tx_receiver: mpsc::Receiver<ConsensusTransaction>,
+        consensus_mode: ConsensusMode,
+    ) {
+        let epoch_store = validator.epoch_store_for_testing().clone();
+        let checkpoint_service = Arc::new(CheckpointServiceNoop {});
+        let counter = IntCounter::new(
+            "skipped_consensus_txns",
+            "Total number of consensus transactions skipped",
+        )
+        .unwrap();
+        while let Some(tx) = tx_receiver.recv().await {
```

Comment:
Not sure what would be the best way to model this, but in practice every consensus commit includes ~1s worth of transactions, so the cost of `process_consensus_transactions` is amortized. Calling the function on every transaction may increase its cost to unrealistic level. Maybe poll every 1s from tx_receiver instead, or poll until a batch size threshold is reached or tx_receiver becomes empty?

Reply:
Added a parameter to group transactions in a commit. Will add arg flags to update it in future PRs (need some more restructuring of the commands first)